File: examples/java-gradle/src/main/java/com/example/impl/FileProductRepository.java
package com.example.impl;

import com.example.interfaces.Repository;
import com.example.models.Product;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.IOException;
import java.util.*;

/**
 * File-based implementation of Product repository.
 */
public class FileProductRepository implements Repository<Product> {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final File file;

    public FileProductRepository(String filePath) {
        this.file = new File(filePath);
    }

    @Override
    public Optional<Product> findById(Long id) {
        return loadAll().stream()
                .filter(p -> id.equals(p.id()))
                .findFirst();
    }

    @Override
    public List<Product> findAll() {
        return loadAll();
    }

    @Override
    public Product save(Product entity) {
        List<Product> products = loadAll();
        Product toSave = entity.id() == null
                ? entity.withId(generateId(products))
                : entity;

        products.removeIf(p -> toSave.id().equals(p.id()));
        products.add(toSave);
        saveAll(products);
        return toSave;
    }

    @Override
    public boolean delete(Long id) {
        List<Product> products = loadAll();
        boolean removed = products.removeIf(p -> id.equals(p.id()));
        if (removed) {
            saveAll(products);
        }
        return removed;
    }

    private List<Product> loadAll() {
        if (!file.exists()) {
            return new ArrayList<>();
        }
        try {
            return MAPPER.readValue(file, new TypeReference<>() {});
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    private void saveAll(List<Product> products) {
        try {
            MAPPER.writeValue(file, products);
        } catch (IOException e) {
            throw new RuntimeException("Failed to save products", e);
        }
    }

    private Long generateId(List<Product> products) {
        return products.stream()
                .map(Product::id)
                .filter(Objects::nonNull)
                .max(Long::compareTo)
                .orElse(0L) + 1;
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/impl/FileProductRepository.kt
package com.example.impl

import com.example.interfaces.Repository
import com.example.models.Product
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * File-based implementation of Product repository.
 */
class FileProductRepository(private val filePath: String) : Repository<Product> {

    private val file = File(filePath)
    private val json = Json { prettyPrint = true }

    override fun findById(id: Long): Product? =
        loadAll().find { it.id == id }

    override fun findAll(): List<Product> = loadAll()

    override fun save(entity: Product): Product {
        val products = loadAll().toMutableList()
        val toSave = if (entity.id == null) {
            entity.withId(generateId(products))
        } else {
            entity
        }

        products.removeAll { it.id == toSave.id }
        products.add(toSave)
        saveAll(products)
        return toSave
    }

    override fun delete(id: Long): Boolean {
        val products = loadAll().toMutableList()
        val removed = products.removeAll { it.id == id }
        if (removed) saveAll(products)
        return removed
    }

    private fun loadAll(): List<Product> {
        if (!file.exists()) return emptyList()
        return try {
            json.decodeFromString<List<Product>>(file.readText())
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun saveAll(products: List<Product>) {
        file.writeText(json.encodeToString(products))
    }

    private fun generateId(products: List<Product>): Long =
        (products.mapNotNull { it.id }.maxOrNull() ?: 0L) + 1
}
